### PR TITLE
Translate from `@types` package to library name

### PR DESCRIPTION
**Pull request by landongrindheim**
Now that we can translate from library to types package (e.g. `jquery` -> `@types/jquery`), we need to be able to translate in the other direction (e.g. `@types/jquery` -> `jquery`).

### CODE REVIEW

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb [hunk 1]
```diff
@@ -11,6 +11,13 @@ class PackageName
           (?<name>[a-z0-9\-~][a-z0-9\-._~]*)         # capture package name
           \z                                         # end of string
       }xi.freeze                                     # multi-line/case-insensitive
+      SCOPED_TYPES_PACKAGE_REGEX = %r{
+          \A                                         # beginning of string
+          @#{DEFINITELY_TYPED_SCOPE}\/               # starts with @types/
+          (?<scope>.+)                               # capture scope
+          __                                         # scoped name separator
+          (?<name>.+)                                # capture name
```
Comment (mattt) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb:19]: We can use a single regex to cover `@types` packages for both scoped and unscoped:

```suggestion
      TYPES_PACKAGE_NAME_REGEX = %r{
          \A                                         # beginning of string
          @#{DEFINITELY_TYPED_SCOPE}\/               # starts with @types/
          ((?<scope>.+)__)?                          # capture scope
          (?<name>.+)                                # capture name
```

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb [hunk 2]
```diff
@@ -11,6 +11,13 @@ class PackageName
...
+      }xi.freeze                                     # multi-line/case-insensitive
```
Comment (mattt) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb:20]: Missing end of string anchor?

```suggestion
          \z                                         # end of string
      }xi.freeze                                     # multi-line/case-insensitive
```

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb [hunk 3]
```diff
@@ -38,6 +45,17 @@ def eql?(other)
...
+      def library_name
+        return self unless types_package?
+
+        @library_name ||=
+          if scoped_types_package?
+            self.class.new(unscoped_types_package_name)
+          else
+            self.class.new(@name.to_s)
+          end
+      end
```
Comment (mattt) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb:57]: If you adopt the `TYPES_PACKAGE_NAME_REGEX` change above, you can avoid helper methods altogether:

```suggestion
      def library_name
        @library_name ||= 
          if match = TYPES_PACKAGE_NAME_REGEX.match(to_s)
            if match[:scope]
              self.class.new("@#{match[:scope]}/#{match[:name]}")
            else
              self.class.new(match[:name].to_s)
            end
          end
      end
```

(Not sure if RuboCop will allow `if` with assignment. You may have to do `match =` and `if match` on separate lines)

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb [hunk 4]
```diff
@@ -55,6 +73,17 @@ def scoped?
...
+        SCOPED_TYPES_PACKAGE_REGEX.match?(to_s)
+      end
+
+      def unscoped_types_package_name
+        match = SCOPED_TYPES_PACKAGE_REGEX.match(to_s)
+        raise InvalidPackageName unless match
+
+        "@#{match[:scope]}/#{match[:name]}"
+      end
+
```
Comment (mattt) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb:86]: ^ From previous suggestion.

```suggestion
```

Comment (landongrindheim) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb]: 👍 All comments addressed together in [14bb063](https://github.com/dependabot/dependabot-core/pull/5025/commits/14bb0637d9a455e903ec498a4291c88c54011571)

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb (hunk 3 above)
Comment (landongrindheim) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb]: 👍 All comments addressed together in [14bb063](https://github.com/dependabot/dependabot-core/pull/5025/commits/14bb0637d9a455e903ec498a4291c88c54011571)

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb (hunk 2 above)
Comment (landongrindheim) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb:20]: 👍 All comments addressed together in [14bb063](https://github.com/dependabot/dependabot-core/pull/5025/commits/14bb0637d9a455e903ec498a4291c88c54011571)

File: npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb (hunk 1 above)
Comment (landongrindheim) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb]: Thanks for the feedback 😄 I'm batching all the changes and adding you as a co-author.

👍 All comments addressed together in [14bb063](https://github.com/dependabot/dependabot-core/pull/5025/commits/14bb0637d9a455e903ec498a4291c88c54011571)

**Review by pavera — APPROVED** [review]
👍  Looks good to me.

**Review by mattt — APPROVED** [review]

**Pull request merged**